rpc: "addpeeraddress tried" return error on failure

0xB10C opened the pull request:
When trying to add an address to the IP address manager tried table, it's first added to the new table and then moved to the tried table. Previously, adding a conflicting address to the address manager's tried table with test-only `addpeeraddress tried=true` RPC would return `{ "success": true }`. However, the address would not be added to the tried table, but would remain in the new table. This caused, e.g., issue #28964.

This is fixed by new returning `{ "success":  false, "error": "..." }` for failed tried table additions. Since the address remaining in the new table can't be removed (the address manager interface does not support removing addresses at the moment and adding this seems to be a bigger effort), an error message is returned. This indicates to a user why the RPC failed and allows accounting for the extra address in the new table. This is done in the functional test for the `getrawaddrman` RPC.

Fixes #28964

0xB10C committed:
rpc: "addpeeraddress tried" return error on failure

When trying to add an address to the IP address manager tried table,
it's first added to the new table and then moved to the tried table.
Previously, adding a conflicting address to the address manager's
tried table with test-only `addpeeraddress tried=true` RPC would
return `{ "success": true }`. However, the address would not be added
to the tried table, but would remain in the new table. This caused,
e.g., issue 28964.

This is fixed by returning `{ "success": false, "error":
"failed-adding-to-tried" }` for failed tried table additions. Since
the address remaining in the new table can't be removed (the address
manager interface does not support removing addresses at the moment
and adding this seems to be a bigger effort), an error message is
returned. This indicates to a user why the RPC failed and allows
accounting for the extra address in the new table.

Also:
To check the number of addresses in each addrman table,
the addrman checks were re-run and the log output of this check
was asserted. Ideally, logs shouldn't be used as an interface
in automated tests. To avoid asserting the logs, use the getaddrmaninfo
and getrawaddrman RPCs (which weren't implemented when the test was added).
Removing the "getnodeaddress" calls would also remove the addrman checks
from the test, which could reduce the test coverage. To avoid this,
these are kept.

pablomartin4btc reviewed (commented):
Concept ACK

jonatack reviewed (commented):
Concept ACK

0xB10C commented on test/functional/rpc_net.py:
```diff
@@ -495,11 +495,20 @@ def check_getrawaddrman_entries(expected):
+                    # new table and then moved to the tried table. If the move to the tried table
+                    # fails, the address still remains in the new table.
+                    if table_name == "tried" and result["error"] == "faild-adding-to-tried":
+                        entry = table_info["entries"][1]
```
> Is a copy the right way, or does it need to be removed from the expected tried table?

Just removing the address from the _expected_ tried table does not work. It's still in the new table. So the question would rather be: "Do we need to retry adding the address to the tried table?". I guess that's debatable. We're only running into this case very rarely when there's a collision - all other runs test that this works. This should be solved with #29007.

> Maybe add steps to reproduce the test code?

Yes, I think we should have that. Especially with typos like https://github.com/bitcoin/bitcoin/pull/28998#discussion_r1416294937. Due to addrman non-determinism, this isn't really doable. Again, #29007 helps here.

0xB10C replied:
Resolving this with the rebase on-top of #29007 as this code was removed in 2cc8ca19f4185490f30a49516c890b2289fbab71. The addresses are now added in [seed_addrman()](https://github.com/bitcoin/bitcoin/pull/29007/commits/2cc8ca19f4185490f30a49516c890b2289fbab71#diff-355d4500a3163e554fc3c3051b21b6bc9a97bcb35b146bff17d5fc1b4e4fd650R44) and ([currently](https://github.com/bitcoin/bitcoin/pull/29007/commits/2cc8ca19f4185490f30a49516c890b2289fbab71#r1521452867)) can't produce a collision.

0xB10C commented:
Maybe get #29007 in first. This allows us to remove the getrawaddrman-test retry logic (possibly do it there). Makes the `getrawaddrman` test simpler.

I've addressed the minor comments here, but I think it makes sense to draft this PR for now until #29007 is in. With #29007 we can test the collision case when `addpeeraddress tried` returns `{"success": false, "error": "failed-adding-to-tried"}` as discussed in https://github.com/bitcoin/bitcoin/pull/28998#discussion_r1417047413.

sr-gi commented on test/functional/rpc_net.py:
```diff
@@ -336,7 +336,7 @@ def test_addpeeraddress(self):
         self.log.debug("Test that adding an already-present tried address to the new and tried tables fails")
         for value in [True, False]:
-            assert_equal(node.addpeeraddress(address="1.2.3.4", tried=value, port=8333), {"success": False})
+            assert_equal(node.addpeeraddress(address="1.2.3.4", tried=value, port=8333), {"success": False, "error": "failed-adding-to-new"})
```
This is a non-exhaustive case now. There are two potential error outcomes that can be triggered on demand, from which only one is being tested:

- a) The address is already in `new` and we are trying to add it again (to any table)
- b) The address is already in `tried` and we are trying to add it again (to any table)

In this case, only `b)` is being tested: we do so both with and without the "try to add to tried" flag, but the reality is that the reason why the function is failing is the entry is already present in `tried`, even when we try to add it to `new`. Given how the addrman works, we cannot really try to add things to `tried` multiple times on demand, given it would need to be added to `new` first and that would fail.

So I think the only way to properly test this would be to add two different addresses twice. e.g:

```python
# Try that the same address cannot be added twice to `new`
assert_equal(node.addpeeraddress(address="1.2.3.5", tried=False, port=8333), {"success": True})
assert_equal(node.addpeeraddress(address="1.2.3.5", tried=False, port=8333), {"success": False, "error": "failed-adding-to-new"})
# Try that the same address cannot be added twice to `tried`
assert_equal(node.addpeeraddress(address="1.2.3.6", tried=True, port=8333), {"success": True})
assert_equal(node.addpeeraddress(address="1.2.3.6", tried=True, port=8333), {"success": False, "error": "failed-adding-to-new"})
```

0xB10C replied:
Done!

sr-gi reviewed (commented):
Approach ACK

It is a pity that `"failed-adding-to-tried"` cannot be tested on demand, however, this at least gets rid of the intermittent failure when a collision is found in `tried`.

I think the tests can be made more exhaustive though, check comment inline.

0xB10C commented:
Thanks for the review @sr-gi! You might want to take a look at #29007 first - this allows us to also test `failed-adding-to-tried`. Once this is merged, I'll continue to work on this PR.

0xB10C committed:
test: remove unused mocktime in test_addpeeraddress

Drops the mocktime added in fa4c6836c9366c3cc575cb386a397840d5f1aa57.
Setting the mocktime in test_addpeeraddress() isn't needed
anymore as it doesn't leak into test_getrawaddrman() anymore
(since 2cc8ca19f4185490f30a49516c890b2289fbab71).

test_getrawaddrman() clear's the addrman and sets it's own
mocktime.

0xB10C commented:
Rebased on #29007, which allows us to check the `failed-adding-to-tried` case. I've also incorporated https://github.com/bitcoin/bitcoin/pull/28998#pullrequestreview-1868852542, and, as I was touching it anyway, replaced the log-assertions `assert_debug_log` on `CheckAddrman: new X, tried Y, total Z started` with `getaddrmaninfo` and `getrawaddrman` calls (which weren't available when the test was added). I've kept the existing `getnodeaddresses` calls, as these re-run the addrman checks.

Also, as I'm touching it already, I've dropped the now unused (since 2cc8ca19f4185490f30a49516c890b2289fbab71) mocktime from `test_addpeeraddress()`. Since #29007, the test does not leak into `test_getrawaddrman()` anymore.

brunoerg commented:
Concept ACK

I had this issue when writing #28869.

stratospher committed:
test: fix test to ensure hidden RPC is present in detailed help

current check to make sure that detailed help for hidden RPC
is displayed won't work because the assertion isn't sufficient.
Even if unknown RPCs are passed, RPC names would still be present
in node.help().

0xB10C commented:
Rebased with #29639 merged due to https://github.com/bitcoin/bitcoin/pull/29639#discussion_r1521951824.

stratospher commented on test/functional/rpc_net.py:
```diff
@@ -340,26 +340,52 @@ def test_addpeeraddress(self):
-            assert_equal(addrs[0]["address"], "1.2.3.4")
-            assert_equal(addrs[0]["port"], 8333)
+        addrman = node.getrawaddrman()
+        assert_equal(len(list(addrman["new"].values())), 1)
```
edc8cb64:
```suggestion
        assert_equal(len(addrman["new"]), 1)
```

stratospher commented on test/functional/rpc_net.py:
```diff
@@ -340,26 +340,52 @@ def test_addpeeraddress(self):
+        self.log.debug("Test that adding a valid address to the new table succeeds")
+        assert_equal(node.addpeeraddress(address="1.0.0.0", tried=False, port=8333), {"success": True})
+        addrman = node.getrawaddrman()
+        assert_equal(len(list(addrman["tried"].values())), 0)
```
edc8cb64: why not?
```suggestion
        assert_equal(len(addrman["tried"]), 0)
```
(left 1 more similar suggestion)

stratospher reviewed (commented):
tested ACK bb0f618.

not touched in this PR but `addpeerinfo` in `test_addpeeraddress` doesn't exist. 😳
if you retouch, do you want to include https://github.com/stratospher/bitcoin/commit/09c849e2f544934b121f5fdf62cb974b67b6b962 since it only affects `addpeeraddress` and `getrawaddrman` and is a simple fix?

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -960,7 +960,8 @@ static RPCHelpMan addpeeraddress()
             {
-                {RPCResult::Type::BOOL, "success", "whether the peer address was successfully added to the address manager"},
+                {RPCResult::Type::BOOL, "success", "whether the peer address was successfully added to the address manager table"},
+                {RPCResult::Type::STR, "error", /*optional=*/true, "error description, if the address could not be added (and success is false)."},
```
In edc8cb64f9aae6939183de276828991243927917: Is it necessary the "(and success is false)"? Isn't it intuitive?

0xB10C commented:
Addressed feedback and included https://github.com/stratospher/bitcoin/commit/09c849e2f544934b121f5fdf62cb974b67b6b962

stratospher reviewed (commented):
reACK 99954f9. 🚀

brunoerg reviewed (approved):
utACK 99954f914f031c80aa53daa367fc049c4c55bdf3

achow101 commented:
ACK 99954f914f031c80aa53daa367fc049c4c55bdf3
